File: LLM-Translations/GPT-4/avatar/Python/C/atcoder_ABC146_D.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

#define MAX_NODES 1000

typedef struct {
    int edges[MAX_NODES];
    int num_edges;
} Node;

typedef struct {
    int a, b;
} Edge;

void bfs(int v, int N, Node G[], int *K, int E[][2]) {
    bool visited[MAX_NODES] = {false};
    int queue[MAX_NODES];
    int front = 0, rear = 0;
    int node2color[MAX_NODES];
    for (int i = 0; i < N; i++) {
        node2color[i] = -1;
    }
    queue[rear++] = v;
    visited[v] = true;
    while (front != rear) {
        int q = queue[front++];
        int color = 0;
        for (int i = 0; i < G[q].num_edges; i++) {
            int nex = G[q].edges[i];
            if (visited[nex]) {
                continue;
            }
            visited[nex] = true;
            color += 1;
            if (color == node2color[q]) {
                color += 1;
            }
            node2color[nex] = color;
            E[nex][0] = q;
            E[nex][1] = color;
            queue[rear++] = nex;
        }
        *K = (*K < color) ? color : *K;
    }
}

int main() {
    int N;
    scanf("%d", &N);
    Node G[MAX_NODES];
    for (int i = 0; i < N; i++) {
        G[i].num_edges = 0;
    }
    int E[MAX_NODES][2];
    for (int i = 0; i < N - 1; i++) {
        int a, b;
        scanf("%d%d", &a, &b);
        a--, b--;
        G[a].edges[G[a].num_edges++] = b;
        G[b].edges[G[b].num_edges++] = a;
    }
    int K = -1;
    bfs(0, N, G, &K, E);
    printf("%d\n", K);
    for (int i = 1; i < N; i++) {
        printf("%d\n", E[i][1]);
    }
    return 0;
}

// End of Code

File: LLM-Translations/GPT-4/avatar/Python/Java/atcoder_ABC146_D.java
import java.util.*;
import java.io.*;

public class atcoder_ABC146_D {
    static int bfs(int v, int N, ArrayList<Deque<Integer>> G, HashMap<SimpleImmutableEntry<Integer, Integer>, Integer> E) {
        int[] visited = new int[N];
        Deque<Integer> queue = new ArrayDeque<>();
        int K = -1;
        int[] node2color = new int[N];
        Arrays.fill(node2color, -1);
        queue.add(v);
        visited[v] = 1;

        while (!queue.isEmpty()) {
            int q = queue.poll();
            int color = 0;
            for (int nex : G.get(q)) {
                if (visited[nex] != 0) {
                    continue;
                }
                visited[nex] = 1;
                color += 1;
                if (color == node2color[q]) {
                    color += 1;
                }
                node2color[nex] = color;
                E.put(new AbstractMap.SimpleImmutableEntry<>(Math.min(q, nex), Math.max(q, nex)), color);
                queue.add(nex);
            }
            K = Math.max(K, color);
        }
        return K;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        ArrayList<Deque<Integer>> G = new ArrayList<>();
        HashMap<AbstractMap.SimpleImmutableEntry<Integer, Integer>, Integer> E = new HashMap<>();
        for (int i = 0; i < N; i++) {
            G.add(new ArrayDeque<>());
        }
        for (int i = 0; i < N - 1; i++) {
            String[] input = br.readLine().split(" ");
            int a = Integer.parseInt(input[0]) - 1;
            int b = Integer.parseInt(input[1]) - 1;
            G.get(a).add(b);
            G.get(b).add(a);
            E.put(new AbstractMap.SimpleImmutableEntry<>(a, b), 0);
        }
        int K = bfs(0, N, G, E);
        System.out.println(K);
        for (int value : E.values()) {
            System.out.println(value);
        }
    }
}
// End of Code.
